hypothesis-python/src/hypothesis/strategies/_internal/core.py

```diff
@@ -2134,9 +2134,20 @@ def draw(self, strategy: SearchStrategy[Ex], label: Any = None) -> Ex:
         self.count += 1
         printer = RepresentationPrinter(context=current_build_context())
         desc = f"Draw {self.count}{'' if label is None else f' ({label})'}: "
+        _global_rand_state = random.getstate()
         result = self.conjecture_data.draw(strategy, observe_as=f"generate:{desc}")
         if TESTCASE_CALLBACKS:
             self.conjecture_data._observability_args[desc] = to_jsonable(result)
+        if _global_rand_state != random.getstate():
+            # raise InvalidDefinition
+            note_deprecation(
+                "Do not use the `random` module inside strategies; instead "
+                "consider  `st.randoms()`, `st.sampled_from()`, etc.  "
+                f"{desc}from {strategy!r}",
+                since="RELEASEDAY",
+                has_codemod=False,
+                stacklevel=1,
+            )
```

Comment:
Maybe consider factoring this out as

```
with checked_random_state():
    result = self.conjecture_data.draw(strategy, observe_as=f"generate:{desc}")
    ...
```
?

That would avoid the duplication w/control.py, plus improve readability in these central functions

Reply:
Yeah, I like it - and we can check a "drew from `st.random_module()` flag there too.